lib/msf/core/post/windows/services.rb

```diff
@@ -175,415 +50,603 @@ def each_service(&block)
-  #   with Display Name). Case sensitive.
-  #
-  # @return [Boolean]
-  #
-  def service_exists?(service)
-    srv_info = service_info(service)
-
-    if srv_info.nil?
-      vprint_error('Unable to enumerate Windows services')
-      return false
-    end
+            next unless [
+              SERVICE_WIN32_OWN_PROCESS,
+              SERVICE_WIN32_OWN_PROCESS_INTERACTIVE,
+              SERVICE_WIN32_SHARE_PROCESS,
+              SERVICE_WIN32_SHARE_PROCESS_INTERACTIVE
+            ].include?(service_type)
 
-    if srv_info && srv_info[:display].empty?
-      return false
-    end
+            yield sk
+          end
 
-    true
-  end
+          keys
+        end
 
-  #
-  # Changes a given service startup mode, name must be provided and the mode.
-  #
-  # Mode is a string with either auto, manual or disable for the
-  # corresponding setting. The name of the service is case sensitive.
-  #
-  # @raise [RuntimeError] if an invalid startup mode is provided in the mode parameter
-  #
-  def service_change_startup(name, mode, server=nil)
-    if mode.is_a? Integer
-      startup_number = mode
-    else
-      case mode.downcase
-        when "boot" then startup_number     = START_TYPE_BOOT
-        when "system" then startup_number   = START_TYPE_SYSTEM
-        when "auto" then startup_number     = START_TYPE_AUTO
-        when "manual" then startup_number   = START_TYPE_MANUAL
-        when "disable" then startup_number  = START_TYPE_DISABLED
-        else
-          raise "Invalid Startup Mode: #{mode}"
-      end
-    end
+        #
+        # List all Windows Services present
+        #
+        # If ExtAPI is available we return the DACL, LOGroup, and Interactive
+        # values otherwise these values are nil
+        #
+        # @return [Array<Hash>] Array of Hashes containing Service details. May contain the following keys:
+        #   * :name
+        #   * :display
+        #   * :pid
+        #   * :status
+        #   * :interactive
+        #
+        # @todo Rewrite to allow operating on a remote host
+        #
+        def service_list
+          return meterpreter_service_list if session.type == 'meterpreter'
+
+          services = []
+          each_service do |s|
+            services << { name: s }
+          end
 
-    if session.railgun
-      begin
-        ret = service_change_config(name, {:starttype => startup_number}, server)
-        return (ret == Error::SUCCESS)
-      rescue Rex::Post::Meterpreter::RequestError => e
-        if server
-          # Cant do remote registry changes at present
-          return false
-        else
-          vprint_error("Request Error #{e} falling back to registry technique")
+          services
         end
-      end
-    end
 
-    servicekey = "HKLM\\SYSTEM\\CurrentControlSet\\Services\\#{name.chomp}"
-    registry_setvaldata(servicekey,"Start",startup_number,"REG_DWORD")
-  end
+        #
+        #
+        # @return
+        def meterpreter_service_list
```

Comment:
This looks like a duplicate of `each_service` given current implementation. Talking this over with the team we'd like to avoid having multiple enumeration functions for this sort of thing. Additionally the naming of this function here implies that it only works on Meterpreter commands whilst this code seems to also work with Shell and Powershell sessions.

Reply:
> This looks like a duplicate of `each_service` given current implementation. Talking this over with the team we'd like to avoid having multiple enumeration functions for this sort of thing.

No

> Additionally the naming of this function here implies that it only works on Meterpreter commands whilst this code seems to also work with Shell and Powershell sessions.

No

Reply:
Fixed with 950bc6da0ee92e2a107dc0ef5893ad01e07c9ac7.

Reply:
Fixed with 950bc6da0ee92e2a107dc0ef5893ad01e07c9ac7